src/librustc_trans/abi.rs

```diff
@@ -407,7 +407,8 @@ impl<'tcx> LayoutExt<'tcx> for TyLayout<'tcx> {
 #[derive(Clone, Copy, PartialEq, Eq, Debug)]
 pub enum CastTarget {
     Uniform(Uniform),
-    Pair(Reg, Reg)
+    Pair(Reg, Reg),
+    ChunkedPrefix { prefix: [RegKind; 8], chunk: Size, total: Size }
```

Comment:
Really starting to wonder if we can turn all 3 variants into one (such that `CastTarget` becomes a `struct`).

If the `prefix` array uses `Option<RegKind>` instead to indicate that not all 8 slots are full, then we can replace `total` with `rest: Uniform` and have it represent strictly what's *after* the prefix.

And then `Pair(a, b)` can become `prefix: [Some(a.kind), None, ..., None], chunk: a.size, rest: Uniform::from(b)`.

Oh and `chunk` makes more sense as `prefix_chunk`.

Reply:
I've attempted to implement this (in separate commits for the moment). It seems to work ok.